### PR TITLE
[RFC] Add test for cycles addon

### PR DESCRIPTION
**Pull request by grische**
**Purpose of the pull request**  
This PR is a quick&dirty CI solution to show the intent of the work which triggers #48

**Description about the pull request**  
It adds a script which tries to lint the cycles addon in Blender together fake-bpy-modules.

**Additional comments** [Optional]  
Just check the failure of the CI to see which errors are remaining for each version of Blender.

### CODE REVIEW

**Comment by nutti** [comment]
I'm not sure why this PR is needed?

**Comment by nutti** [comment]
As I mentioned in slack channel, I will advice two things.

* E1111 and E1011 error is difficult to handle in fake-bpy-module, so we should add these error to `IGNORED_PYLINT_ERRORS`.
* E1101 should be remained, but I recommend the CI should pass because this error sometimes needs lots of time to resolve.

**Comment by grische** [comment]
> As I mentioned in slack channel, I will advice two things.
> 
> * E1111 and E1011 error is difficult to handle in fake-bpy-module, so we should add these error to `IGNORED_PYLINT_ERRORS`.
> * E1101 should be remained, but I recommend the CI should pass because this error sometimes needs lots of time to resolve.

I tried disabling the errors as you asked (E1011 does not exist), but when I disabled all errors, the "Should-be-failing pylint test" did not fail anymore. That would have gave the test no meaning.

But I found a more elegant way to ignore errors and only explicitly ignore `E1111` (next to `E0601` and `E0604`), by creating a [pylintrc](https://github.com/PyCQA/pylint/blob/master/pylintrc) file:
```rc
[MESSAGES CONTROL]
disable=C,R,W,E0601,E0602,E1111

[TYPECHECK]
ignored-modules=_cycles
ignored-classes=CyclesPreferences
```

This ignores the `CyclesPreferences` class as well as `_cycles` module.

I hope you like the improved version.

**Comment by nutti** [comment]
@grische 

Cool!
It looks greater than before.
From the output of this test (v2.83), I think `bl_ui` and `bl_operators` modules are missing.
We need to think about this...

```bash
Running test again with fake-bpy-module:
************* Module addon.ui
intern/cycles/blender/addon/ui.py:21:0: E0401: Unable to import 'bl_ui.utils' (import-error)
intern/cycles/blender/addon/ui.py:25:0: E0401: Unable to import 'bl_ui.properties_grease_pencil_common' (import-error)
intern/cycles/blender/addon/ui.py:2363:4: E1101: Class 'RENDER_PT_context' has no 'append' member (no-member)
intern/cycles/blender/addon/ui.py:2364:4: E1101: Class 'VIEW3D_HT_header' has no 'append' member (no-member)
intern/cycles/blender/addon/ui.py:2376:4: E1101: Class 'RENDER_PT_context' has no 'remove' member (no-member)
intern/cycles/blender/addon/ui.py:2377:4: E1101: Class 'VIEW3D_HT_header' has no 'remove' member (no-member)
************* Module addon.operators
intern/cycles/blender/addon/operators.py:133:30: E1120: No value for argument 'msgctxt' in function call (no-value-for-parameter)
************* Module addon.presets
intern/cycles/blender/addon/presets.py:19:0: E0401: Unable to import 'bl_operators.presets' (import-error)

------------------------------------------------------------------
Your code has been rated at 9.85/10 (previous run: 9.45/10, +0.40)
```

**Comment by grische** [comment]
rebased on top of master

**Comment by nutti** [comment]
@grische 

#48 is now resolved.
I think newest master branch solves errors raised from pylint.
Could you try it?

**Comment by grische** [comment]
@nutti I created a new commit which tries to work around some of the issues we are encountering:

**Issues for 2.7x:**
1. ```js
   E1101: Module 'bpy.types' has no 'CYCLES_MT_sampling_presets' member (no-member)
   ``` 
    There is this odd code `bpy.types.CYCLES_MT_sampling_presets...` which should probably be simply `CYCLES_MT_sampling_presets` (as it is defined in the same files). Replacing the code to fix the error.
1. ```js
   E1101: Class 'IMAGE_UV_sculpt' has no 'COMPAT_ENGINES' member (no-member)
   ```
   As this is a known pylint issue: https://github.com/PyCQA/pylint/issues/801 , I am simply injecting a "pylint disable" into the source to ignore it

**Issues for 2.8x:**
1. ```js
   E1120: No value for argument 'layout' in unbound method call (no-value-for-parameter)
   ```
    I don't understand enough of this code, but it seems that this layout function [should have two arguments](https://github.com/blender/blender/blob/607d745a79e0c1f79fdd026625be4169e5530001/release/scripts/startup/bl_ui/utils.py#L32-L38) `(cls, layout)`, so I am simply "fixing" the cycles code before running the test.
1. ```js
   E1120: No value for argument 'msgctxt' in function call (no-value-for-parameter)
   ```
    As far as I understand, the argument `msgctxt` of `pgettext_tip` should be optional. It is used across several places of the code as `tip_` and always has only one argument. Is this a bug of fake bpy?

**Comment by nutti** [comment]
@grische 

Thanks for your investigation.

> E1101: Module 'bpy.types' has no 'CYCLES_MT_sampling_presets' member (no-member)

Blender register classes to `bpy.types` module by calling `register_class`.
So, this is valid when Blender is launched.
I think we can ignore this error because most of developers will not use CYCLE_MT_sampling_presets directly.

> E1101: Class 'IMAGE_UV_sculpt' has no 'COMPAT_ENGINES' member (no-member)

Your workaround looks good to me. Thanks!

> E1120: No value for argument 'layout' in unbound method call (no-value-for-parameter)

I think this method comes from `Panel` class which is a base class of `CYCLES_PT_sampling_presets`.
I think this error is raised when calling the classmethod implicitly derived from base class.
I'm not sure why pylint could not handle this case, but your workaround is enough for disabling this weird error.

> E1120: No value for argument 'msgctxt' in function call (no-value-for-parameter)

This is a bug of fake-bpy-module.
I have already fixed it. (https://github.com/nutti/fake-bpy-module/commit/cfc1749cfa1fb92abb8ca62a3daa79d73cbb163e)

File: tests/pylint_cycles.sh [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+git fetch --quiet --depth 1 origin "${remote_git_ref}"
+git checkout -f "${remote_git_ref}"
+
+# Generate and print pylintrc
+pylintrcpath="${temp_venv}/pylintrc"
+
+create_pylintrc "${pylintrcpath}"
+echo "Testing with the following pylintrc:"
+cat "${pylintrcpath}"
+echo
```
Comment (nutti) [inline tests/pylint_cycles.sh:173]: Question: Why echo commands with empty argument are exist after some commands?

Comment (grische) [reply tests/pylint_cycles.sh:173]: @nutti it is a nice way to generate an empty line to separate outputs

File: tests/pylint_cycles.sh [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+# Fixing addon code to workaround some quirks
+workaround_quirks ${version}
+echo
+
+# Expect failure before fake-bpy-modul is installed, otherwise following test has no meaning
```
[the remark this answers is not in the document]
Comment (grische) [reply tests/pylint_cycles.sh]: fixed

File: tests/pylint_cycles.sh [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        # There is an issue with CYCLES_PT_sampling_presets.draw_panel_header() requiring two arguments
+        # but only one is delivered.
+        echo "Fixing cycles quirk: \".draw_panel_header(self.layout)\""
+        sed -i 's/.draw_panel_header(self.layout)/.draw_panel_header(self, layout)/' intern/cycles/blender/addon/ui.py
+    fi
+
+    if [[ $version =~ ^2.7[89]$ ]]; then
+        # Grische: I am not sure why, but instead of using the class directly, they are inferering the class
+        # from bpy.types which fake-bpy-module does not provide
+        echo "Fixing cycles quirk: \".bpy.types.CYCLES_MT_[a-z]*_presets\""
```
[the remark this answers is not in the document]
Comment (grische) [reply tests/pylint_cycles.sh:122]: fixed

File: tests/pylint_cycles.sh [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+# check if PYTHON_BIN binary is availble
+if ! command -v "${PYTHON_BIN}" > /dev/null; then
+    echo "Error: Cannot find ${PYTHON_BIN} binary."
+    exit 1
+fi
+python_bin=$(command -v "${PYTHON_BIN}")
+
+# check if python version meets our requirements
+IFS=" " read -r -a python_version <<< "$(${python_bin} -c 'import sys; print(sys.version_info[:])' | tr -d '(),')"
+if [ "${python_version[0]}" -lt 3 ]; then
```
[the remark this answers is not in the document]
Comment (grische) [reply tests/pylint_cycles.sh]: fixed

File: tests/pylint_cycles.sh [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    echo "ignored-modules=_cycles" >> "${pylintrc}"
+    echo "ignored-classes=CyclesPreferences" >> "${pylintrc}"
+}
+
+function workaround_quirks() {
+    local version=$1
+
+    if [[ $version =~ ^2.8[0-9]$ ]]; then
+        # There is an issue with CYCLES_PT_sampling_presets.draw_panel_header() requiring two arguments
+        # but only one is delivered.
```
[the remark this answers is not in the document]
Comment (grische) [reply tests/pylint_cycles.sh]: fixed

**Comment by grische** [comment]
@nutti updated the PR and turns from draft into proper PR. Feel free to merge if you think it is OK.

File: tests/pylint_cycles.sh (hunk 1 above)
Comment (nutti) [reply tests/pylint_cycles.sh:173]: Thanks for your explanation.
I understood.

**Pull request merged**